### PR TITLE
Add zoom support for scaled components

### PR DESCRIPTION
**Pull request by staradayev**
In case of using *CSS* style
`transform: scale()` for container or any another parent element for zoom option,
DragSelect has wrong selector rendering position and selected events has been fired also incorrect.
 
It happens because _transform: scale()_ property changes visual size of presentation at page nested elements, but in fact pixels size stays the same.

So, there in this PR, the *zoom* property has been added, where you can pass your scale factor (zoom) options used at parent.

This zoom is used to compensate scaled presentation dimensions.

image of DragSelect with scale

### CODE REVIEW

**Comment by ThibaultJanBeyer** [comment]
Wow, thanks a lot @staradayev for your contribution! I'm currently in holidays and will have a look at your PR next week. Just wanted to let you know that this sounds like a good addition and that seeing you wanting to contribute makes me happy. Thank you!

File: src/DragSelect.js [hunk 1]
```diff
@@ -238,7 +241,7 @@ class DragSelect {
       this._prevSelected = [];
     } // #9
 
-    this.checkIfInsideSelection(true); // reset selection if no multiselectionkeypressed
+    this.checkIfInsideSelection(true, this.zoom); // reset selection if no multiselectionkeypressed
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: no need to pass it as props if we use the Object approach and check for zoom using `this.zoom`

File: src/DragSelect.js [hunk 2]
```diff
@@ -300,8 +303,7 @@ class DragSelect {
       event.preventDefault();
 
     if (this._isRightClick(event)) return;
-    if (this._isScrollbarClick(event, this.area)) return;
-
+    if (this._isScrollbarClick(event, this.area, this.zoom)) return;
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: no need to pass it as props if we use the Object approach and check for zoom using `this.zoom`

File: src/DragSelect.js [hunk 3]
```diff
@@ -316,7 +318,7 @@ class DragSelect {
 
     // move element on location
     this._getStartingPositions(event);
-    this.checkIfInsideSelection(true);
+    this.checkIfInsideSelection(true, this.zoom);
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: no need to pass it as props if we use the Object approach and check for zoom using `this.zoom`

File: src/DragSelect.js [hunk 4]
```diff
@@ -368,7 +370,8 @@ class DragSelect {
   _getStartingPositions(event) {
     this._initialCursorPos = this._newCursorPos = this._getCursorPos(
       event,
-      this.area
+      this.area,
+      this.zoom
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: no need to pass it as props if we use the Object approach and check for zoom using `this.zoom`

File: src/DragSelect.js [hunk 5]
```diff
@@ -405,7 +408,7 @@ class DragSelect {
 
     // move element on location
     this._updatePos(this.selector, selectorPos);
-    this.checkIfInsideSelection();
+    this.checkIfInsideSelection(null, this.zoom);
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: no need to pass it as props if we use the Object approach and check for zoom using `this.zoom`

File: src/DragSelect.js [hunk 6]
```diff
@@ -418,7 +421,7 @@ class DragSelect {
    * @private
    */
   _getPosition(event) {
-    var cursorPosNew = this._getCursorPos(event, this.area);
+    var cursorPosNew = this._getCursorPos(event, this.area, this.zoom);
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: no need to pass it as props if we use the Object approach and check for zoom using `this.zoom`

File: src/DragSelect.js [hunk 7]
```diff
@@ -510,20 +513,19 @@ class DragSelect {
    * @param {boolean} [force] forces through. Handles first clicks and accessibility. Here is user is clicking directly onto some element at start, (contrary to later hovers) we can assume that he really wants to select/deselect that item.
    * @return {boolean}
    */
-  checkIfInsideSelection(force) {
+  checkIfInsideSelection(force, zoom) {
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: no need to pass it as props if we use the Object approach and check for zoom using `this.zoom` in the method

File: src/DragSelect.js [hunk 8]
```diff
@@ -510,20 +513,19 @@ class DragSelect {
...
-        y: this.selector.getBoundingClientRect().top + scroll.y,
-        x: this.selector.getBoundingClientRect().left + scroll.x,
+        y: (this.selector.getBoundingClientRect().top)/this.zoom + scroll.y,
+        x: (this.selector.getBoundingClientRect().left)/this.zoom + scroll.x,
         h: this.selector.offsetHeight,
         w: this.selector.offsetWidth
       };
-
-      if (this._isElementTouching(selectable, selectionRect, scroll)) {
+      if (this._isElementTouching(selectable, selectionRect, scroll, zoom)) {
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: no need to pass it as props if we use the Object approach and check for zoom using `this.zoom` in the method

File: src/DragSelect.js [hunk 9]
```diff
@@ -647,13 +649,13 @@ class DragSelect {
    * @return {boolean}
    * @private
    */
-  _isElementTouching(element, selectionRect, scroll) {
+  _isElementTouching(element, selectionRect, scroll, zoom) {
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: no need to pass it as props if we use the Object approach and check for zoom using `this.zoom` in the method

File: src/DragSelect.js [hunk 10]
```diff
@@ -647,13 +649,13 @@ class DragSelect {
...
     const rect = element.getBoundingClientRect();
     const elementRect = {
-      y: rect.top + scroll.y,
-      x: rect.left + scroll.x,
-      h: rect.height,
-      w: rect.width
+      y: (rect.top/zoom + scroll.y),
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: use object `this.zoom` instead here. Also, brackets are not nescessary.

File: src/DragSelect.js [hunk 11]
```diff
@@ -647,13 +649,13 @@ class DragSelect {
...
+      x: (rect.left/zoom + scroll.x),
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: use object `this.zoom` instead here. Also, no need for the brackets.

File: src/DragSelect.js [hunk 12]
```diff
@@ -647,13 +649,13 @@ class DragSelect {
...
+      h: rect.height/zoom,
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: use object `this.zoom` instead here.

File: src/DragSelect.js [hunk 13]
```diff
@@ -647,13 +649,13 @@ class DragSelect {
...
+      w: rect.width/zoom
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: use object `this.zoom` instead here.

File: src/DragSelect.js [hunk 14]
```diff
@@ -690,7 +692,7 @@ class DragSelect {
    * @private
    */
   _autoScroll(event) {
-    var edge = this.isCursorNearEdge(event, this.area);
+    var edge = this.isCursorNearEdge(event, this.area, this.zoom);
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: no need to pass it as props if we use the Object approach and check for zoom using this.zoom in the method

File: src/DragSelect.js [hunk 15]
```diff
@@ -716,8 +718,8 @@ class DragSelect {
    * @param {(HTMLElement|SVGElement)} area the area.
    * @return {('top'|'bottom'|'left'|'right'|false)}
    */
-  isCursorNearEdge(event, area) {
-    var cursorPosition = this._getCursorPos(event, area);
+  isCursorNearEdge(event, area, zoom) {
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: no need to pass it as props if we use the Object approach and check for zoom using this.zoom in the method

File: src/DragSelect.js [hunk 16]
```diff
@@ -716,8 +718,8 @@ class DragSelect {
...
+    var cursorPosition = this._getCursorPos(event, area, zoom);
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: no need to pass it as props if we use the Object approach and check for zoom using this.zoom in the method

File: src/DragSelect.js [hunk 17]
```diff
@@ -755,7 +757,7 @@ class DragSelect {
    * @param {boolean} [withCallback] - whether or not the callback should be called
    */
   reset(event, withCallback) {
-    this._previousCursorPos = this._getCursorPos(event, this.area);
+    this._previousCursorPos = this._getCursorPos(event, this.area, this.zoom);
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: no need to pass it as props if we use the Object approach and check for zoom using this.zoom in the method

File: src/DragSelect.js [hunk 18]
```diff
@@ -837,7 +840,7 @@ class DragSelect {
     if (!event) return { x: 0, y: 0 };
 
     var area = _area || (_area !== false && this.area);
-    var pos = this._getCursorPos(event, area);
+    var pos = this._getCursorPos(event, area, this.zoom);
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: no need to pass it as props if we use the Object approach and check for zoom using this.zoom in the method

File: src/DragSelect.js [hunk 19]
```diff
@@ -1039,13 +1042,13 @@ class DragSelect {
    * @return {boolean}
    * @private
    */
-  _isScrollbarClick(event, area) {
-    const cPos = this._getCursorPos(event, area);
+  _isScrollbarClick(event, area, zoom) {
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: no need to pass it as props if we use the Object approach and check for zoom using this.zoom in the method

File: src/DragSelect.js [hunk 20]
```diff
@@ -1039,13 +1042,13 @@ class DragSelect {
...
+    const cPos = this._getCursorPos(event, area, zoom);
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: no need to pass it as props if we use the Object approach and check for zoom using this.zoom in the method

File: src/DragSelect.js [hunk 21]
```diff
@@ -1099,10 +1102,11 @@ class DragSelect {
    * Use the public .getCursorPos() from outside, it’s more flexible
    * @param {Object} [event]
    * @param {(HTMLElement|SVGElement)} area – containing area / document if none
+   * @param {number} [zoom] - zoom factor
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: no need to pass it as props if we use the Object approach and check for zoom using this.zoom in the method

File: src/DragSelect.js [hunk 22]
```diff
@@ -1120,12 +1124,11 @@ class DragSelect {
...
 
     var areaRect = this.getAreaRect(area || document);
-    var docScroll = this.getScroll(); // needed when document is scrollable but area is not
-
+    var docScroll = this.getScroll(null); // needed when document is scrollable but area is not
     return {
       // if it’s constrained in an area the area should be substracted calculate
-      x: cPos.x - areaRect.left - docScroll.x,
-      y: cPos.y - areaRect.top - docScroll.y
+      x: (cPos.x - areaRect.left - docScroll.x)/zoom,
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: use `this.zoom` instead. Also, are the brackets nescessary?

File: src/DragSelect.js [hunk 23]
```diff
@@ -1120,12 +1124,11 @@ class DragSelect {
...
+      y: (cPos.y - areaRect.top - docScroll.y)/zoom
```
Comment (ThibaultJanBeyer) [inline src/DragSelect.js]: use `this.zoom` instead. Also, are the brackets nescessary?

File: src/DragSelect.js [hunk 24]
```diff
@@ -68,6 +68,7 @@ class DragSelect {
    * @param {object} options - The options object.
    * @param {HTMLElement | SVGElement | Document} [options.area=document] area in which you can drag. If not provided it will be the whole document
    * @param {number} [options.autoScrollSpeed=1] Speed in which the area scrolls while selecting (if available). Unit is pixel per movement. Default = 1
+   * @param {number} [options.zoom=1] Zoom (if available). Unit sczle soom. Default = 1
```
[the remark this answers is not in the document]
Comment (staradayev) [reply src/DragSelect.js]: +

File: src/DragSelect.js [hunk 25]
```diff
@@ -1120,12 +1124,11 @@ class DragSelect {
     };
 
     var areaRect = this.getAreaRect(area || document);
-    var docScroll = this.getScroll(); // needed when document is scrollable but area is not
-
+    var docScroll = this.getScroll(null); // needed when document is scrollable but area is not
```
[the remark this answers is not in the document]
Comment (staradayev) [reply src/DragSelect.js]: No... Fixed.

File: src/DragSelect.js [hunk 26]
```diff
@@ -1039,13 +1042,13 @@ class DragSelect {
...
-  _isScrollbarClick(event, area) {
-    const cPos = this._getCursorPos(event, area);
+  _isScrollbarClick(event, area, zoom) {
+    const cPos = this._getCursorPos(event, area, zoom);
     const areaRect = this.getAreaRect(area);
     const border = area.computedBorder || 0;
 
-    if (areaRect.width + border <= cPos.x) return true;
-    if (areaRect.height + border <= cPos.y) return true;
+    if ((areaRect.width + border) <= cPos.x) return true;
```
[the remark this answers is not in the document]
Comment (staradayev) [reply src/DragSelect.js]: No, fixed

File: src/DragSelect.js [hunk 27]
```diff
@@ -1099,10 +1102,11 @@ class DragSelect {
    * Use the public .getCursorPos() from outside, it’s more flexible
    * @param {Object} [event]
    * @param {(HTMLElement|SVGElement)} area – containing area / document if none
+   * @param {number} [zoom] - zoom factor
    * @return {{x: number, y: number}} cursor X/Y
    * @private
    */
-  _getCursorPos(event, area) {
+  _getCursorPos(event, area, zoom) {
```
[the remark this answers is not in the document]
Comment (staradayev) [reply src/DragSelect.js]: Done

File: src/DragSelect.js [hunk 28]
```diff
@@ -771,6 +773,7 @@ class DragSelect {
     this.selector.style.width = '0';
     this.selector.style.height = '0';
     this.selector.style.display = 'none';
+
```
[the remark this answers is not in the document]
Comment (staradayev) [reply src/DragSelect.js]: Removed

File: __tests__/quicktest.html [hunk 29]
```diff
@@ -152,23 +152,51 @@
...
   <script src="../docs/DragSelect.js"></script>
   <!-- <script type="module">
     import DragSelect from "../docs/ds.es6m.min.js"; -->
   <script>
-  var ds = new DragSelect({
-    selectables: document.getElementsByClassName('item'),
-    area: document.getElementById('container'),
-    callback: function(e) { console.log("selected", e); }
-  });
+  var zoom = 1.5;
```
[the remark this answers is not in the document]
Comment (staradayev) [reply __tests__/quicktest.html]: So yes. I've reverted back `quicktest` to the original.
Same time, I've created zoom html, spec and css files in `__tests__/functional/`.
Tests for selecting elements on zoomed container at beginning and scrolled end automated now. Please review

File: src/DragSelect.js (hunk 1 above)
Comment (staradayev) [reply src/DragSelect.js]: Done.

File: src/DragSelect.js (hunk 2 above)
Comment (staradayev) [reply src/DragSelect.js]: Done.

File: src/DragSelect.js (hunk 3 above)
Comment (staradayev) [reply src/DragSelect.js]: Done.

File: src/DragSelect.js (hunk 4 above)
Comment (staradayev) [reply src/DragSelect.js]: Done.

File: src/DragSelect.js (hunk 5 above)
Comment (staradayev) [reply src/DragSelect.js]: Done.

File: src/DragSelect.js (hunk 6 above)
Comment (staradayev) [reply src/DragSelect.js]: Done.

File: src/DragSelect.js (hunk 19 above)
Comment (staradayev) [reply src/DragSelect.js]: Done.

File: src/DragSelect.js (hunk 20 above)
Comment (staradayev) [reply src/DragSelect.js]: Done.

File: src/DragSelect.js (hunk 21 above)
Comment (staradayev) [reply src/DragSelect.js]: Done.

File: src/DragSelect.js (hunk 22 above)
Comment (staradayev) [reply src/DragSelect.js]: No, the brackets are not necessary. Removed.

File: src/DragSelect.js (hunk 23 above)
Comment (staradayev) [reply src/DragSelect.js]: Done.

File: src/DragSelect.js (hunk 7 above)
Comment (staradayev) [reply src/DragSelect.js]: Done

File: src/DragSelect.js (hunk 8 above)
Comment (staradayev) [reply src/DragSelect.js]: Done

File: src/DragSelect.js (hunk 9 above)
Comment (staradayev) [reply src/DragSelect.js]: Done

File: src/DragSelect.js (hunk 10 above)
Comment (staradayev) [reply src/DragSelect.js]: Done

File: src/DragSelect.js (hunk 11 above)
Comment (staradayev) [reply src/DragSelect.js]: Done

File: src/DragSelect.js (hunk 12 above)
Comment (staradayev) [reply src/DragSelect.js]: Done

File: src/DragSelect.js (hunk 13 above)
Comment (staradayev) [reply src/DragSelect.js]: Done

File: src/DragSelect.js (hunk 14 above)
Comment (staradayev) [reply src/DragSelect.js]: Done

File: src/DragSelect.js (hunk 15 above)
Comment (staradayev) [reply src/DragSelect.js]: Done

File: src/DragSelect.js (hunk 16 above)
Comment (staradayev) [reply src/DragSelect.js]: Done

File: src/DragSelect.js (hunk 17 above)
Comment (staradayev) [reply src/DragSelect.js]: Done

File: src/DragSelect.js (hunk 18 above)
Comment (staradayev) [reply src/DragSelect.js]: Done

**Comment by ThibaultJanBeyer** [comment]
Thank you! I’ll update the readme accordingly and publish a new version soon

**Comment by staradayev** [comment]
Thank you!

**Comment by staradayev** [comment]
Hey!

@ThibaultJanBeyer can you please update **npm** package also? Because it's 1.12.2 currently available latest there. 
We want to update a package in the project with latest changes there.

Thank you!

**Comment by ThibaultJanBeyer** [comment]
Oh thanks for the reminder! Done.

**Pull request merged**